### PR TITLE
[storyteller] adding a grep wrapper with predefined scenarios

### PR DESCRIPTION
**Pull request by yxieca**
**- What I did**

Introduce storyteller tool to grep syslog (or any other logs) for
a story line of a certain scenario.

Defined scenarios are:
- reboot  : device reboot related events.
- service : service start/stop events.
- link    : link up/down events.
- lag     : LAG up/down events.
- bgp     : BGP config change events.
- crash   : process/service crash events.

Unmatched cateory is used as grepping regex directly.

Signed-off-by: Ying Xie <ying.xie@microsoft.com>

**- How to verify it**
Used this tool in some recent investigations.

### CODE REVIEW

File: scripts/storyteller [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    Utility to help analyze log for certain sequence of events.
+    e.g.: reboot (including warm/fast reboot), interface flapping, etc.
+'''
+
+import argparse
+import subprocess
+
+regex_dict = {
+                'bgp'       : 'bgpcfgd',
+                'crash'     : 'what\|unexpected exception\|notify_OA_about_syncd_exception\|SIG',
```
Comment (prsunny) [inline scripts/storyteller:14]: should we also add "not expected" ?

Comment (yxieca) [reply scripts/storyteller:14]: Glad to get some inputs :-)

Comment (yxieca) [reply scripts/storyteller]: Done. Thanks Prince!

**Comment by yxieca** [comment]
retest this please

File: scripts/storyteller [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
```
Comment (jleveque) [inline scripts/storyteller:1]: sonic-utilities is now only installed as a Python 3 package, so this shebang *should* be fine. However, until we completely remove Python 2 from the image, I suggest changing this to `#!/usr/bin/env python3`, just to be sure.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: scripts/storyteller (hunk 2 above)
Comment (yxieca) [reply scripts/storyteller]: Thanks! Done!

File: scripts/storyteller [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                'crash'     : 'what\|unexpected exception\|notify_OA_about_syncd_exception\|SIG\|not expected',
+                'interface' : 'updatePortOperStatus\|Configure .* to',
+                'lag'       : 'link becomes\|addLag',
+                'reboot'    : 'BOOT\|rc.local\|old_config\|minigraph.xml\|Rebooting\|reboot\|executeOperationsOnAsic\|getAsicView\|dumpVidToAsicOperatioId',
+                'service'   : 'Starting\|Stopping\|Started\|Stopped',
+             }
+
+
+def exec_cmd(cmd):
+    out = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True)
```
Comment (jleveque) [inline scripts/storyteller:23]: Suggest adding `text=True` parameter (this causes the return values to be string type in Python 3), then you can remove the `.decode(encoding='UTF-8')`s on line 25.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: scripts/storyteller (hunk 3 above)
Comment (yxieca) [reply scripts/storyteller]: This is so much better! Thanks!

File: scripts/storyteller [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                        type=int, required=False, default=0)
+    parser.add_argument('-C', '--context', help='Show N lines before and after match',
+                        type=int, required=False, default=0)
+
+    args = parser.parse_args()
+
+    log = args.log
+    reg = build_regex(args.category)
+
+
```
Comment (jleveque) [inline scripts/storyteller:81]: Suggest to delete one of these blank lines, as it is inside the function. Two blank lines should be reserved for between functions. Sorry for the very iterative review.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: scripts/storyteller (hunk 4 above)
Comment (yxieca) [reply scripts/storyteller]: sure. done.

**Review by jleveque — APPROVED** [review]

**Pull request merged**